File: Scripts/fetch-posters.js
/**
 * Boxoffy — TMDB Poster Fetcher
 * Usage: node scripts/fetch-posters.js
 *
 * Reads src/data/films.json, queries TMDB for each film,
 * writes back posterUrl field. Safe to re-run — skips films
 * that already have a posterUrl.
 */

const fs   = require('fs');
const path = require('path');

const TOKEN = 'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI2YWM1OTY0NzEyZTA0NGRmMmJjMmFiYzFlMTFlZGMyYyIsIm5iZiI6MTc3Mzg4NTMxMS42MzIsInN1YiI6IjY5YmI1NzdmYjgyMzJhNzc5MjIxZWZjOCIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.QxWkqL_4_3PfTnIVgcZYE0zlXQd9M2y5QL-oWNwq6dE';
const BASE  = 'https://api.themoviedb.org/3';
const IMG   = 'https://image.tmdb.org/t/p/w185';

// Language → TMDB language code for better matching
const LANG_MAP = {
  'Hindi':     'hi',
  'Telugu':    'te',
  'Tamil':     'ta',
  'Kannada':   'kn',
  'Malayalam': 'ml',
  'Marathi':   'mr',
  'Manipuri':  'mni',
  'Hollywood': 'en',
};

// Manual overrides for titles that TMDB indexes differently
const OVERRIDES = {
  'K.G.F: Chapter 2':            'KGF Chapter 2',
  'K.G.F: Chapter 1':            'KGF Chapter 1',
  'Ponniyin Selvan I':            'Ponniyin Selvan Part 1',
  'Ponniyin Selvan II':           'Ponniyin Selvan Part 2',
  'Brahmastra Part One':          'Brahmastra Part One: Shiva',
  'Salaar: Part 1':               'Salaar: Part 1 – Ceasefire',
  'Spider-Man: No Way Home (Hindi)': 'Spider-Man: No Way Home',
  'Mufasa: The Lion King (Hindi)': 'Mufasa: The Lion King',
  'Captain America: Brave New World': 'Captain America: Brave New World',
  'MSVP (Mana ShankaraVaraprasad Garu)': 'Mana ShankaraVaraprasad Garu',
  'Dhurandhar: The Revenge':      'Dhurandhar 2',
  'Chatha Pacha: The Ring of Rowdies': 'Chatha Pacha',
  'Krantijyoti Vidyalay Marathi Madhyam': 'Krantijyoti',
  'NTR – Neel':                   'Dragon',
  'Ustaad Bhagat Singh':          'Ustaad Bhagat Singh',
  'Anaganaga Oka Raju':           'Anaganaga Oka Raju',
  'Nari Nari Naduma Murari':      'Nari Nari Naduma Murari',
  'Mana ShankaraVaraprasad Garu': 'Mana ShankaraVaraprasad Garu',
};

async function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function searchTMDB(title, language, year) {
  const query    = OVERRIDES[title] || title;
  const langCode = LANG_MAP[language] || 'hi';

  // Try 1: with language + year
  const url1 = `${BASE}/search/movie?query=${encodeURIComponent(query)}&language=en-US&page=1&include_adult=false${year ? `&year=${year}` : ''}`;
  const r1 = await fetch(url1, {
    headers: { Authorization: `Bearer ${TOKEN}`, accept: 'application/json' }
  });
  const d1 = await r1.json();
  if (d1.results && d1.results.length > 0 && d1.results[0].poster_path) {
    return d1.results[0].poster_path;
  }

  // Try 2: without year
  const url2 = `${BASE}/search/movie?query=${encodeURIComponent(query)}&language=en-US&page=1&include_adult=false`;
  const r2 = await fetch(url2, {
    headers: { Authorization: `Bearer ${TOKEN}`, accept: 'application/json' }
  });
  const d2 = await r2.json();
  if (d2.results && d2.results.length > 0 && d2.results[0].poster_path) {
    return d2.results[0].poster_path;
  }

  return null;
}

async function main() {
  const filePath = path.join(__dirname, '../src/data/films.json');
  const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  let updated = 0;
  let skipped = 0;
  let failed  = 0;

  for (const year of Object.keys(data)) {
    for (const film of data[year]) {
      if (film.posterUrl) {
        skipped++;
        continue;
      }

      // Extract year from releaseDate if possible
      const yearMatch = (film.releaseDate || '').match(/\b(20\d{2})\b/);
      const releaseYear = yearMatch ? yearMatch[1] : year;

      try {
        const posterPath = await searchTMDB(film.title, film.language, releaseYear);
        if (posterPath) {
          film.posterUrl = `${IMG}${posterPath}`;
          console.log(`✓  ${film.title} → ${film.posterUrl}`);
          updated++;
        } else {
          console.log(`✗  ${film.title} — no poster found`);
          failed++;
        }
      } catch (e) {
        console.log(`✗  ${film.title} — error: ${e.message}`);
        failed++;
      }

      // TMDB rate limit: 40 req / 10 sec → ~250ms between calls is safe
      await sleep(260);
    }
  }

  fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
  console.log(`\nDone. Updated: ${updated} | Skipped (already had poster): ${skipped} | Failed: ${failed}`);
  console.log('films.json written.');
}

main().catch(console.error);

File: fetch_posters.py
"""
BOXOFFY — TMDB Poster Fetcher
Run locally: python fetch_posters.py
Reads:  src/data/films.json
Writes: src/data/films.json (updated with posterUrl for each film)

Requires: pip install requests
"""

import json, time, requests, re, os

# ── CONFIG ────────────────────────────────────────────────────
BEARER   = "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI2YWM1OTY0NzEyZTA0NGRmMmJjMmFiYzFlMTFlZGMyYyIsIm5iZiI6MTc3Mzg4NTMxMS42MzIsInN1YiI6IjY5YmI1NzdmYjgyMzJhNzc5MjIxZWZjOCIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.QxWkqL_4_3PfTnIVgcZYE0zlXQd9M2y5QL-oWNwq6dE"
BASE     = "https://api.themoviedb.org/3"
IMG_BASE = "https://image.tmdb.org/t/p/w300"   # w300 = good quality, reasonable size
DATA     = "src/data/films.json"
DELAY    = 0.26  # seconds between requests — stays under TMDB rate limit (40/10s)

HEADERS = {
    "Authorization": f"Bearer {BEARER}",
    "accept": "application/json",
}

# ── LANGUAGE CODE MAP ─────────────────────────────────────────
LANG_MAP = {
    "Hindi":      "hi",
    "Telugu":     "te",
    "Tamil":      "ta",
    "Malayalam":  "ml",
    "Kannada":    "kn",
    "Hollywood":  "en",
    "Marathi":    "mr",
    "Punjabi":    "pa",
    "Bengali":    "bn",
}

def clean_year(date_str):
    """Extract 4-digit year from releaseDate string"""
    m = re.search(r'(\d{4})', str(date_str))
    return m.group(1) if m else None

def search_tmdb(title, year, lang_code):
    """Search TMDB for a film, return poster_path or None"""
    # Try with year first
    try:
        r = requests.get(f"{BASE}/search/movie", headers=HEADERS, params={
            "query":               title,
            "year":                year,
            "language":            "en-US",
            "include_adult":       False,
        }, timeout=8)
        r.raise_for_status()
        results = r.json().get("results", [])
        if results:
            return results[0].get("poster_path")
    except Exception:
        pass

    # Fallback: search without year
    try:
        r = requests.get(f"{BASE}/search/movie", headers=HEADERS, params={
            "query":   title,
            "language":"en-US",
        }, timeout=8)
        r.raise_for_status()
        results = r.json().get("results", [])
        if results:
            return results[0].get("poster_path")
    except Exception:
        pass

    return None

def run():
    # Load films.json
    with open(DATA, encoding="utf-8") as f:
        films = json.load(f)

    total    = sum(len(v) for v in films.values() if isinstance(v, list))
    updated  = 0
    already  = 0
    failed   = 0
    count    = 0

    print(f"=== BOXOFFY TMDB POSTER FETCHER ===")
    print(f"Films to process: {total}")
    print(f"Estimated time:   ~{round(total * DELAY / 60, 1)} minutes\n")

    for year in sorted(films.keys()):
        fl = films[year]
        if not isinstance(fl, list):
            continue

        for film in fl:
            count += 1
            title = film.get("title", "")

            # Skip if already has a poster
            if film.get("posterUrl") and "tmdb.org" in str(film.get("posterUrl", "")):
                already += 1
                continue

            lang      = film.get("language", "Hindi")
            lang_code = LANG_MAP.get(lang, "hi")
            yr        = clean_year(film.get("releaseDate", year))

            poster_path = search_tmdb(title, yr, lang_code)
            time.sleep(DELAY)

            if poster_path:
                film["posterUrl"] = f"{IMG_BASE}{poster_path}"
                updated += 1
                print(f"  [{count:>3}/{total}] ✅  {title[:50]}")
            else:
                failed += 1
                print(f"  [{count:>3}/{total}] ❌  {title[:50]} (not found)")

            # Save progress every 50 films in case of interruption
            if count % 50 == 0:
                with open(DATA, "w", encoding="utf-8") as f:
                    json.dump(films, f, ensure_ascii=False, indent=2)
                print(f"\n  💾 Progress saved ({count}/{total})\n")

    # Final save
    with open(DATA, "w", encoding="utf-8") as f:
        json.dump(films, f, ensure_ascii=False, indent=2)

    print(f"\n=== DONE ===")
    print(f"Updated:        {updated}")
    print(f"Already had:    {already}")
    print(f"Not found:      {failed}")
    print(f"Total films:    {total}")
    print(f"\nNext step: python generate_film_pages.py")
    print(f"Then:       git add -A && git commit -m 'feat: TMDB posters' && npm run build && vercel --prod")

if __name__ == "__main__":
    run()
